sync/pending_polling.go

```diff
@@ -0,0 +1,673 @@
+			// ignore the incoming pending if it has fewer transactions than the one we already have
+			return false, nil
+		}
+	} else if !errors.Is(err, ErrPendingBlockNotFound) {
+		return false, err
+	}
+
+	s.pendingData.Store(utils.HeapPtr[core.PendingData](p))
+
+	return true, nil
+}
+
+// UpdatePreLatestAttachment updates (or clears) the PreLatest attachment of the currently stored
+// pre_confirmed at the given blockNumber by atomically swapping the stored pointer.
+// Returns true if the store was updated, false if no matching pre_confirmed is stored
+// or the attachment was already equal.
+func (s *Synchronizer) UpdatePreLatestAttachment(blockNumber uint64, preLatest *core.PreLatest) bool {
+	curPtr := s.pendingData.Load()
+	if curPtr == nil || *curPtr == nil {
+		return false
+	}
+
+	cur := *curPtr
+	pc, ok := cur.(*core.PreConfirmed)
+	if !ok {
+		// A pending is stored or another type; nothing to do.
+		return false
+	}
+	if pc.Block == nil || pc.Block.Number != blockNumber {
+		// Different height stored; do not touch.
+		return false
+	}
+
+	if pc.PreLatest == preLatest {
+		// No change.
+		return false
+	}
+
+	// Clone and update attachment on the clone.
+	next := pc.Copy()
+	next.WithPreLatest(preLatest)
+
+	return s.pendingData.CompareAndSwap(curPtr, utils.HeapPtr[core.PendingData](next))
+}
+
+// StorePreConfirmed stores a pre_confirmed block given that it is for the next height.
+// If an equal-number block with >= txCount already exists, we do not overwrite it,
+// but we allow updating the PreLatest attachment in-place via a CAS swap.
+func (s *Synchronizer) StorePreConfirmed(p *core.PreConfirmed) (bool, error) {
+	if err := blockchain.CheckBlockVersion(p.GetBlock().ProtocolVersion); err != nil {
+		return false, err
+	}
+
+	head, err := s.blockchain.HeadsHeader()
+	if err != nil {
+		if !errors.Is(err, db.ErrKeyNotFound) {
+			return false, err
+		}
+		head = nil
+	}
+
+	if !p.Validate(head) {
+		return false, errors.New("store pre_confirmed not valid for parent")
+	}
+
+	existingPtr := s.pendingData.Load()
+	if existingPtr == nil || *existingPtr == nil {
+		return s.pendingData.CompareAndSwap(
+			existingPtr,
+			utils.HeapPtr[core.PendingData](p),
+		), nil
+	}
+
+	existing := *existingPtr
+	if !existing.Validate(head) {
+		return s.pendingData.CompareAndSwap(
+			existingPtr,
+			utils.HeapPtr[core.PendingData](p),
+		), nil
+	}
+
+	existingB := existing.GetBlock()
+	incomingB := p.GetBlock()
+
+	// If same number and existing has >= txCount, keep existing but update attachment if needed.
+	if existingB.Number == incomingB.Number &&
+		existingB.TransactionCount >= incomingB.TransactionCount {
+		// Try to update attachment to match p.PreLatest (can be nil to clear)
+		_ = s.UpdatePreLatestAttachment(incomingB.Number, p.PreLatest)
+		return false, nil
+	}
+
+	// Otherwise, swap in the better pre_confirmed.
+	return s.pendingData.CompareAndSwap(
+		existingPtr,
+		utils.HeapPtr[core.PendingData](p),
+	), nil
```

Comment:
I think if it's easier to keep track if we can reduce the number of returns. For example, something like:
```
	shouldPreserveCurrentPreConfirmed := existingPtr != nil && *existingPtr == nil && !(*existingPtr).Validate(head) && ...
	if shouldPreserveCurrentPreConfirmed {
		_ = s.UpdatePreLatestAttachment(incomingB.Number, p.PreLatest)
		return false, nil
	}
	return s.pendingData.CompareAndSwap(
		existingPtr,
		utils.HeapPtr[core.PendingData](p),
	), nil
```